pkg/config/configobserver/configmapobserver.go

```diff
@@ -0,0 +1,133 @@
+package configobserver
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/openshift/insights-operator/pkg/config"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+)
+
+const (
+	insightsConfigMapName = "insights-config"
+)
+
+type ConfigMapInformer interface {
+	factory.Controller
+	// Config provides actual Insights configuration values from the "insights-config" configmap
+	Config() *config.InsightsConfiguration
+	// ConfigChanged notifies all the listeners that the content of the "insights-config" configmap has changed
+	ConfigChanged() (<-chan struct{}, func())
+}
+
+// ConfigMapObserver is a controller for "insights-config" config map
+// in the "openshift-insights" namespace.
+type ConfigMapObserver struct {
+	factory.Controller
+	lock           sync.Mutex
+	kubeCli        *kubernetes.Clientset
+	insightsConfig *config.InsightsConfiguration
+	listeners      map[chan struct{}]struct{}
+}
+
+func NewConfigMapObserver(kubeConfig *rest.Config,
+	eventRecorder events.Recorder,
+	kubeInformer v1helpers.KubeInformersForNamespaces) (ConfigMapInformer, error) {
+	cmInformer := kubeInformer.InformersFor("openshift-insights").Core().V1().ConfigMaps().Informer()
+	kubeClient, err := kubernetes.NewForConfig(kubeConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	ctrl := &ConfigMapObserver{
+		kubeCli:        kubeClient,
+		insightsConfig: nil,
+		listeners:      make(map[chan struct{}]struct{}),
+	}
+	factoryCtrl := factory.New().WithInformers(cmInformer).
+		WithSync(ctrl.sync).
+		ResyncEvery(10*time.Minute).
+		ToController("ConfigController", eventRecorder)
+
+	ctrl.Controller = factoryCtrl
+	return ctrl, nil
+}
+
+// sync is called by the informer with every config map update
+func (c *ConfigMapObserver) sync(ctx context.Context, _ factory.SyncContext) error {
+	cm, err := getConfigMap(ctx, c.kubeCli)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+		// config map doesn't exist so clear the config, notify and return
+		klog.Info(err)
+		c.insightsConfig = nil
+		c.notifyListeners()
+		return nil
+	}
+	insightsConfig, err := readConfigAndDecode(cm)
+	if err != nil {
+		return err
+	}
+	// do not notify listeners on resync
+	if c.insightsConfig != insightsConfig {
+		c.insightsConfig = insightsConfig
+		c.notifyListeners()
+	}
+	return nil
+}
+
+func (c *ConfigMapObserver) notifyListeners() {
+	for ch := range c.listeners {
+		if ch == nil {
+			continue
+		}
+		ch <- struct{}{}
+	}
+}
+
+func (c *ConfigMapObserver) Config() *config.InsightsConfiguration {
+	c.lock.Lock()
```

Comment:
I think that if we use `sync.RWMutex` we can use `RLock` and `RUnlock` here... it would be probably better.

Reply:
How about this one?